sycl/include/sycl/nd_item.hpp

```diff
@@ -38,6 +41,502 @@ namespace ext::oneapi::experimental {
+    size_t LinId = 0;
+    id<Dimensions> Index = get_local_id();
+    range<Dimensions> Extent = get_local_range();
+    if (1 == Dimensions) {
+      LinId = Index[0];
+    } else if (2 == Dimensions) {
+      LinId = Index[0] * Extent[1] + Index[1];
+    } else {
+      LinId =
+          Index[0] * Extent[1] * Extent[2] + Index[1] * Extent[2] + Index[2];
+    }
+    __SYCL_ASSUME_INT(LinId);
+    return LinId;
+  }
+
+  group<Dimensions> get_group() const {
+    // TODO: ideally Group object should be stateless and have a contructor with
+    // no arguments.
+    return detail::Builder::createGroup(get_global_range(), get_local_range(),
+                                        get_group_range(), get_group_id());
+  }
+
+  sub_group get_sub_group() const { return sub_group(); }
+
+  size_t __SYCL_ALWAYS_INLINE get_group(int Dimension) const {
+    size_t Id = get_group_id()[Dimension];
+    __SYCL_ASSUME_INT(Id);
+    return Id;
+  }
+
+  size_t __SYCL_ALWAYS_INLINE get_group_linear_id() const {
+    size_t LinId = 0;
+    id<Dimensions> Index = get_group_id();
+    range<Dimensions> Extent = get_group_range();
+    if (1 == Dimensions) {
+      LinId = Index[0];
+    } else if (2 == Dimensions) {
+      LinId = Index[0] * Extent[1] + Index[1];
+    } else {
+      LinId =
+          Index[0] * Extent[1] * Extent[2] + Index[1] * Extent[2] + Index[2];
+    }
+    __SYCL_ASSUME_INT(LinId);
+    return LinId;
+  }
+
+  range<Dimensions> get_group_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initNumWorkgroups<Dimensions, range<Dimensions>>();
+#else
+    throw sycl::exception(sycl::make_error_code(sycl::errc::invalid),
+                          "nd_item methods can't be invoked on the host");
+    return {};
+#endif
+  }
+
+  size_t __SYCL_ALWAYS_INLINE get_group_range(int Dimension) const {
+    size_t Range = get_group_range()[Dimension];
+    __SYCL_ASSUME_INT(Range);
+    return Range;
+  }
+
+  range<Dimensions> get_global_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalSize<Dimensions, range<Dimensions>>();
+#else
+    throw sycl::exception(sycl::make_error_code(sycl::errc::invalid),
+                          "nd_item methods can't be invoked on the host");
+    return {};
+#endif
+  }
+
+  size_t get_global_range(int Dimension) const {
+    size_t Val = get_global_range()[Dimension];
+    __SYCL_ASSUME_INT(Val);
+    return Val;
+  }
+
+  range<Dimensions> get_local_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initWorkgroupSize<Dimensions, range<Dimensions>>();
+#else
+    throw sycl::exception(sycl::make_error_code(sycl::errc::invalid),
+                          "nd_item methods can't be invoked on the host");
+    return {};
+#endif
+  }
+
+  size_t get_local_range(int Dimension) const {
+    size_t Id = get_local_range()[Dimension];
+    __SYCL_ASSUME_INT(Id);
+    return Id;
+  }
+
+  __SYCL2020_DEPRECATED("offsets are deprecated in SYCL 2020")
```

Comment:
I wonder if it makes sense to have deprecated functions under the `__INTEL_PREVIEW_BREAKING_CHANGES` guard.  Won't these functions be removed at the next major release?  If that is the case, I think it might make sense to omit the deprecated functions in this mode because that would be a more accurate preview of the next major release.

Reply:
I suggest removing deprecated methods in a separate PR. If there are going to be removed with in next major release, we should put existing deprecated methods of `nd_item` under `#ifndef __INTEL_PREVIEW_BREAKING_CHANGES`. Otherwise, we will have different sets of methods for stateless and stateful implementations.